src/exact.jl

```diff
@@ -0,0 +1,319 @@
+    end
+
+    # add constraints for source
+    for (i, μi) in zip(axes(xmat, 1), μ) # eachrow(xmat) is not available on Julia 1.0
+        f = MOI.ScalarAffineFunction(
+            [MOI.ScalarAffineTerm(one(μi), xi) for xi in view(xmat, i, :)], zero(μi)
+        )
+        MOI.add_constraint(model, f, MOI.EqualTo(μi))
+    end
+
+    # add constraints for target
+    for (i, νi) in zip(axes(xmat, 2), ν) # eachcol(xmat) is not available on Julia 1.0
+        f = MOI.ScalarAffineFunction(
+            [MOI.ScalarAffineTerm(one(νi), xi) for xi in view(xmat, :, i)], zero(νi)
+        )
+        MOI.add_constraint(model, f, MOI.EqualTo(νi))
+    end
+
+    # compute optimal solution
+    MOI.optimize!(model)
+    status = MOI.get(model, MOI.TerminationStatus())
+    status === MOI.OPTIMAL || error("failed to compute optimal transport plan: ", status)
+    p = MOI.get(model, MOI.VariablePrimal(), x)
+    γ = reshape(p, nμ, nν)
+
+    return γ
+end
+
+"""
+    emd2(μ, ν, C, optimizer; plan=nothing)
+
+Compute the optimal transport cost (a scalar) for the Monge-Kantorovich problem with source
+histogram `μ`, target histogram `ν`, and cost matrix `C` of size `(length(μ), length(ν))`
+which is given by
+```math
+\\inf_{γ ∈ Π(μ, ν)} \\langle γ, C \\rangle.
+```
+
+The corresponding linear programming problem is solved with the user-provided `optimizer`.
+Possible choices are `Tulip.Optimizer()` and `Clp.Optimizer()` in the `Tulip` and `Clp`
+packages, respectively.
+
+A pre-computed optimal transport `plan` may be provided.
+"""
+function emd2(μ, ν, C, optimizer; plan=nothing)
+    γ = if plan === nothing
+        # compute optimal transport plan
+        emd(μ, ν, C, optimizer)
+    else
+        # check dimensions
+        size(C) == (length(μ), length(ν)) ||
+            error("cost matrix `C` must be of size `(length(μ), length(ν))`")
+        size(plan) == size(C) || error(
+            "optimal transport plan `plan` and cost matrix `C` must be of the same size",
+        )
+        plan
+    end
+    return dot(γ, C)
+end
+
+###################################
+# Semidiscrete and continuous 1D OT
+###################################
+
+"""
+    ot_plan(c, μ::ContinuousUnivariateDistribution, ν::UnivariateDistribution)
+
+Compute the optimal transport plan for the Monge-Kantorovich problem with univariate
+distributions `μ` and `ν` as source and target marginals and cost function `c` of
+the form ``c(x, y) = h(|x - y|)`` where ``h`` is a convex function.
+
+In this setting, the optimal transport plan is the Monge map
+```math
+T = F_\\nu^{-1} \\circ F_\\mu
+```
+where ``F_\\mu`` is the cumulative distribution function of `μ` and ``F_\\nu^{-1}`` is the
+quantile function of `ν`.
+
+See also: [`ot_cost`](@ref), [`emd`](@ref)
+"""
+function ot_plan(c, μ::ContinuousUnivariateDistribution, ν::UnivariateDistribution)
+    # Use T instead of γ to indicate that this is a Monge map.
+    T(x) = quantile(ν, cdf(μ, x))
+    return T
+end
```

Comment:
I just realized that we can improve this in a way that the code could actually handle distributions mu with a mix of continuous and discrete. The way it's now, we need mu to be continuous. Since the other function for 1D OT only deals with finite discrete measures, we are leaving it out measure such as Poisson (which is discrete, but with infinite support).
Although, I guess we can add it in another PR and merge this now, since an efficient way of computing 1D OT is very useful.

Reply:
Such cases require a new/different function since the plan is not a Monge map anymore. I agree that this should be done in a separate PR.